Word Prompter in Rap Session Screen

kensonhui opened the pull request:
#related issues

kensonhui committed:
Added word generator

-Button that triggers next word
-Button that fetches random words
-Counter of words remaining
-Potato Code!

kensonhui committed:
Adjusted Landing Page

-Link to Trainer
-Linker from trainer to home
-I messed with tailwind config and now some colours work

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,55 @@
+            }
+        })
+      };
+      const [words, setWords] = useState({
```
why is this hook used inside changeWordHandler?

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,55 @@
+import "tailwindcss/tailwind.css";
+
+const WordPrompt = () => {
+    const changeWordHandler = () => {
```
The indentation seems a bit off here, `useState` and `fetchWords` seem to be within `changeWordHandler` if when they aren't

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,55 @@
+      }
+
+      return (
+        <div className="flex flex-col m-10">
```
check/fix indentation

kensonhui replied:
Thanks, indentation should be fixed now

kensonhui committed:
Adding Chakra, Update Tailwind

Installation of Charka Provider to App
Updated tailwindcss version, 2.1,1. version did not support ':checked'

kensonhui committed:
Fixed Indentation, Added Prerap screen

kensonhui commented:
My bad I realized I added too many commits. In this branch, I also added Chakra UI and updated TailwindCSS's version. I didn't realize I committed in the same branch, next time I will put separate features in different branches.

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,57 @@
+    });
+  };
+  const [words, setWords] = useState({
+    currentWord: "Start",
```
Not sure how I feel about this, this seems kinda hacky. Since we will have a start screen later on, I don't think we will need to set the first word to "start"

gzcharleszhang replied:
Instead of having a separate `currentWord` state, can't we just try to display the first word in `nextWords`? This way you don't have to keep shifting the word into `currentWord`, and you just need to check if `nextWords` is empty when displaying the value. This also separates UI logic from your `changeWordHandler`

kensonhui replied:
You're right, I've combined both into the same state as you suggested

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,57 @@
+const WordPrompt = () => {
+  const changeWordHandler = () => {
+    setWords((lastWords) => {
+      if (lastWords.nextWords.length != 0) {
```
If you follow my other comment, this if statement/branching shouldn't be necessary

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,57 @@
+        });
+      })
+      .catch((error) => {
+        console.log(error);
```
we shouldn't merge console logs into the main branch, you could just not catch anything for now, if we catch a bug we will fix it

gzcharleszhang replied:
Normally, if there's an error, you'd wanna display some error to the user or log it in the database. But this is outside the scope of this PR

kensonhui replied:
Okay, got it. I've removed the catch, and I'll handle it in a later issue.

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,57 @@
+          currentWord: lastWords.nextWords.shift(),
+        };
+      } else {
+        return {
```
maybe we can just do a `fetchWords` here, instead of letting the list go empty

kensonhui replied:
I've set it now so that when the word count goes below 3, it'll call fetchWords

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,57 @@
+      .then((response) => response.json())
+      .then((data) => {
+        setWords((lastWords) => {
+          lastWords.nextWords.push(...data);
```
Be careful of the order of pushing/popping words here. Since we are popping words from the end of the array in `changeWordHandler`, shouldn't we add words to the front of the array? Something like
```
return [ ...data, ...lastWords]
```
Assuming that you followed my above comment and `lastWords` is just the `nextWords` array in the state

gzcharleszhang commented on pages/trainer/index.js:
```diff
@@ -0,0 +1,26 @@
+        <a>Home</a>
+      </Link>
+      <button 
+      onClick={startStopHandler}
```
fix indentation please

kensonhui replied:
Fixed

kensonhui committed:
State Implementation Change, Auto-Fetch Words

-currentWord and nextWord are now combined as the same state
-Session now begins by fetching words instead of only "Start"
-Words are fetched automatically when list of words goes below 3 words
-Removed randomwords api error catch
-Fixed indentation

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,46 @@
+  const changeWordHandler = () => {
+    setWords((lastWords) => {
+      if (lastWords.length <= 3) {
+        fetchWords();
```
do you need this within `setWords`? I think we could do this check before we call `setWords`

gzcharleszhang commented on components/wordPrompt.js:
```diff
@@ -0,0 +1,46 @@
+      .then((data) => {
+        setWords((lastWords) => {
+          lastWords.push(...data);
+          return [...lastWords];
```
you might be able to do `[...lastWords, ...data]` here, then you don't need to do `lastWords.push(...data)`

gzcharleszhang reviewed (approved):
looks good after those 2 comments, make sure to test that it works before merging, gj!

kensonhui committed:
Fixes to pushing to array

-Changed from using lastWords.push(data) to returning [...lastWords, ...data]
-check length of word before calling setWords
